src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java

```diff
@@ -1527,34 +1327,25 @@ static MethodHandle varargsArray(Class<?> arrayType, int nargs) {
         MethodHandle cache[] = Makers.TYPED_COLLECTORS.get(elemType);
         MethodHandle mh = nargs < cache.length ? cache[nargs] : null;
         if (mh != null)  return mh;
-        if (nargs == 0) {
-            Object example = java.lang.reflect.Array.newInstance(arrayType.getComponentType(), 0);
-            mh = MethodHandles.constant(arrayType, example);
-        } else if (elemType.isPrimitive()) {
-            MethodHandle builder = getConstantHandle(MH_fillNewArray);
-            MethodHandle producer = buildArrayProducer(arrayType);
-            mh = buildVarargsArray(builder, producer, nargs);
-        } else {
-            Class<? extends Object[]> objArrayType = arrayType.asSubclass(Object[].class);
-            Object[] example = Arrays.copyOf(NO_ARGS_ARRAY, 0, objArrayType);
-            MethodHandle builder = getConstantHandle(MH_fillNewTypedArray).bindTo(example);
-            MethodHandle producer = getConstantHandle(MH_arrayIdentity); // must be weakly typed
-            mh = buildVarargsArray(builder, producer, nargs);
-        }
-        mh = mh.asType(MethodType.methodType(arrayType, Collections.<Class<?>>nCopies(nargs, elemType)));
-        mh = makeIntrinsic(mh, Intrinsic.NEW_ARRAY);
+        mh = makeCollector(arrayType, nargs);
         assert(assertCorrectArity(mh, nargs));
         if (nargs < cache.length)
             cache[nargs] = mh;
         return mh;
     }
 
-    private static MethodHandle buildArrayProducer(Class<?> arrayType) {
-        Class<?> elemType = arrayType.getComponentType();
-        assert(elemType.isPrimitive());
-        return getConstantHandle(MH_copyAsPrimitiveArray).bindTo(Wrapper.forPrimitiveType(elemType));
+    private static boolean assertCorrectArity(MethodHandle mh, int arity) {
+        assert(mh.type().parameterCount() == arity) : "arity != "+arity+": "+mh;
+        return true;
     }
 
+    // Array identity function (used as getConstantHandle(MH_arrayIdentity)).
```

Comment:
If this is no longer used, remove it.

Reply:
Yes, it looks like this can be removed. Good catch!